src/sbvr-api/uri-parser.ts

```diff
@@ -0,0 +1,225 @@
+})()
+
+export const metadataEndpoints = [ '$metadata', '$serviceroot' ]
+
+const notBadRequestOrParsingError = (e: Error) => {
+	return !((e instanceof BadRequestError) || (e instanceof ParsingError))
+}
+
+export const parseOData = (b: UnparsedRequest) => {
+	return Promise.try<ODataRequest | ODataRequest[]>(() => {
+		if (b._isChangeSet && b.changeSet != null) {
+			const env = new Map<ODataRequest['id'], ODataRequest>()
+			// We sort the CS set once, we must assure that requests which reference
+			// other requests in the changeset are placed last. Once they are sorted
+			// Map will guarantee retrival of results in insertion order
+			const sortedCS = _.sortBy(b.changeSet, (el) => el.url[0] !== '/')
+			return Promise.reduce(sortedCS, parseODataChangeset, env)
+			.then((env) => Array.from(env.values()) as ODataRequest[])
+		} else {
+			const { url, apiRoot } = splitApiRoot(b.url)
+			const odata = odataParser.matchAll(url, 'Process')
+
+			return {
+				method: b.method as SupportedMethod,
+				vocabulary: apiRoot,
+				resourceName: odata.tree.resource,
+				odataBinds: odata.binds,
+				odataQuery: odata.tree,
+				values: b.data,
+				custom: {},
+				_defer: false,
+			}
+		}
+	}).catch(SyntaxError, () => {
+		throw new BadRequestError(`Malformed url: '${b.url}'`)
+	}).catch(notBadRequestOrParsingError, (e) => {
+		console.error('Failed to parse url: ', b.method, b.url, e, e.stack)
+		throw new ParsingError(`Failed to parse url: '${b.url}'`)
+	})
+}
+
+const parseODataChangeset = (env: Map<ODataRequest['id'], ODataRequest>, b: UnparsedRequest) => {
+	const contentId: ODataRequest['id'] = mustExtractHeader(b, 'content-id')
+
+	if (env.has(contentId)) {
+		throw new BadRequestError('Content-Id must be unique inside a changeset')
+	}
+
+	let defer: boolean
+	let odata
+	let apiRoot: string
+	let url
+
+	if (b.url[0] === '/') {
+		({ url, apiRoot } = splitApiRoot(b.url))
+		odata = odataParser.matchAll(url, 'Process')
+		defer = false
+	} else {
+		url = b.url
+		odata = odataParser.matchAll(url, 'Process')
+		const { bind } = odata.tree.resource
+		const [ , id ] = odata.binds[bind]
+		// Use reference to collect information
+		const ref = env.get(id)
+		if (_.isUndefined(ref)) {
+			throw new BadRequestError('Content-Id refers to a non existent resource')
+		}
+		apiRoot = ref.vocabulary
+		// Update resource with actual resourceName
+		odata.tree.resource = ref.resourceName
+		defer = true
+	}
+
+	const parseResult: ODataRequest = {
+		method: b.method as SupportedMethod,
+		vocabulary: apiRoot,
+		resourceName: odata.tree.resource,
+		odataBinds: odata.binds,
+		odataQuery: odata.tree,
+		values: b.data,
+		custom: {},
+		id: contentId,
+		_defer: defer,
+	}
+	env.set(contentId, parseResult)
+	return env
+}
+
+const splitApiRoot = (url: string) => {
+	let urlParts = url.split('/')
+	const apiRoot = urlParts[1]
+	if (apiRoot == null || odata2AbstractSQL[apiRoot] == null) {
+		throw new ParsingError(`No such api root: ${apiRoot}`)
```

Comment:
Not much point logging `No such api root: null` if `apiRoot` is null -- maybe break in two separate ifs that throw the same error with different messages?

Reply:
Agreed, seeing that error in logs is still going to leave questions open, might as well minimise that.